doc/api/addons.md

```diff
@@ -251,6 +251,74 @@ down. If necessary, such hooks can be removed using
 `RemoveEnvironmentCleanupHook()` before they are run, which has the same
 signature.
 
+#### ABI declaration
+
+Node.js is available from a number of sources besides the [official
+distribution][]. Since the various versions of Node.js are configured
+differently at build time, the resulting runtime ABI may be different. For
+example, version 10 of Node.js as shipped by Debian GNU/Linux may have a
+different ABI than version 10 of Node.js as available from the official
+distribution.
+
+The Node.js ABI consists of various different, independent parts, such as V8,
+openssl, libuv, and others. Native addons may use some, all, or even just one of
+these independent parts of the Node.js ABI. Thus, when Node.js is tasked with
+loading an addon, at which time it needs to determine whether the addon is ABI-
+compatible, it needs ABI information provided by the addon. The addon normally
+provides this information as a single number (`NODE_MODULE_VERSION`) which is
+stored inside the addon and which is compared against the value present in the
+running Node.js process at addon load time.
+
+Since `NODE_MODULE_VERSION` reflects only the Node.js major version against
+which the addon was built, it may match the running Node.js process even though
+some of the independent parts of the ABI are mismatched. To address this problem
+the addon may optionally declare which portions of the Node.js ABI it uses by
+invoking the `NODE_MODULE_DECLARE_ABI` macro. Any portions of the ABI included
+as a parameter to the macro will be checked during addon load in addition to
+`NODE_MODULE_VERSION` in order to ensure that all ABIs declared by the addon
+have the version as requested by the addon. Node.js assumes that ABIs not
+included in the invocation of the `NODE_MODULE_DECLARE_ABI` macro are not used
+by the addon.
```

Comment:
We should document what happens if there is a mismatch.

Reply:
@mcollina I moved the documentation to the section titled "Linking to Node.js' own dependencies" and added documentation about the mismatch scenario.